dnf5-plugins/needs_restarting_plugin/needs_restarting.cpp

```diff
@@ -0,0 +1,306 @@
+    if (stat("/proc/1", &proc_1_stat) == 0) {
+        proc_1_boot_time = proc_1_stat.st_mtime;
+    }
+
+    time_t uptime_boot_time = 0;
+    std::ifstream uptime_stream{"/proc/uptime"};
+    if (uptime_stream.is_open()) {
+        double uptime = 0;
+        uptime_stream >> uptime;
+        if (uptime > 0) {
+            uptime_boot_time = std::time(nullptr) - static_cast<time_t>(uptime);
+        }
+    }
+
+    return std::max(proc_1_boot_time, uptime_boot_time);
+}
+
+libdnf5::rpm::PackageSet recursive_dependencies(
+    const libdnf5::rpm::Package & package, libdnf5::rpm::PackageQuery & installed) {
+    libdnf5::rpm::PackageSet dependencies{package.get_base()};
+    dependencies.add(package);
+
+    std::vector<libdnf5::rpm::Package> stack;
+    stack.emplace_back(package);
+
+    while (!stack.empty()) {
+        const auto & current = stack.back();
+        stack.pop_back();
+
+        libdnf5::rpm::PackageQuery query{installed};
+        query.filter_provides(current.get_requires());
+
+        for (const auto & dependency : query) {
+            if (!dependencies.contains(dependency)) {
+                stack.emplace_back(dependency);
+            }
+        }
+
+        dependencies |= query;
+    }
+
+    return dependencies;
+}
+
+void NeedsRestartingCommand::system_needs_restarting(Context & ctx) {
+    const auto boot_time = get_boot_time();
+
+    libdnf5::rpm::PackageQuery base_query{ctx.base};
+
+    libdnf5::rpm::PackageQuery installed{base_query};
+    installed.filter_installed();
+
+    libdnf5::rpm::PackageQuery reboot_suggested{installed};
+    reboot_suggested.filter_reboot_suggested();
+
+    std::vector<libdnf5::rpm::Package> need_reboot = {};
+    for (const auto & pkg : reboot_suggested) {
+        if (pkg.get_install_time() > static_cast<unsigned long long>(boot_time)) {
+            need_reboot.push_back(pkg);
+        }
+    }
+
+    if (need_reboot.empty()) {
+        std::cout << "No core libraries or services have been updated since boot-up.\n"
+                  << "Reboot should not be necessary.\n";
+    } else {
+        std::cout << "Core libraries or services have been updated since boot-up:\n";
+        for (const auto & pkg : need_reboot) {
+            std::cout << "  * " << pkg.get_name() << '\n';
+        }
+        std::cout << "\nReboot is required to fully utilize these updates.\n"
+                  << "More information: https://access.redhat.com/solutions/27943\n";
+        throw libdnf5::cli::SilentCommandExitError(1);
+    }
+}
+
+void NeedsRestartingCommand::services_need_restarting(Context & ctx) {
+    std::unique_ptr<sdbus::IConnection> connection;
+    try {
+        connection = sdbus::createSystemBusConnection();
+    } catch (const sdbus::Error & ex) {
+        throw libdnf5::cli::CommandExitError(1, M_("Couldn't connect to D-Bus: {}"), ex.what());
+    }
+
+    auto systemd_proxy = sdbus::createProxy("org.freedesktop.systemd1", "/org/freedesktop/systemd1");
```

Comment:
I would make the constants for these D-Bus paths, they are already used on multiple places here.